_posts/linda_xiang/2016-12-07-Cloud-data-transfer.md

```diff
@@ -0,0 +1,143 @@
+   image:  linda_xiang/data-transfer/github.png# optional
+header: 
+   version: small
+   title: Software Engineering Blog
+   image: header-logo-crop.png
+   icon: icon-blog
+---
+
+## The task
+
+The Pan-Cancer Analysis of Whole Genomes
+([PCAWG](https://dcc.icgc.org/pcawg)) project is a collaborative effort by
+the International Cancer Genome Consortium (ICGC) to identify common
+mutation pattens from over 2,800 cancer whole genomes with an emphasize in
+non-coding regions. Large amount of data were generated and stored in
+multiple GNOS repositories, core dataset includes:
+
+* 703.89 TB of whole genome uniformly aligned (by BWA-MEM) sample-level BAM
+* 695.59 GB of Sanger/DKFZ/Broad/Muse variant calling VCF
+* 24.52 TB of RNA-Seq TopHat2/Star aligned BAM
+
+The ICGC Data Coordination Centre ([DCC](https://dcc.icgc.org/)) was tasked to transfer this
+dataset to cloud based storage systems, specifically, OpenStack Ceph
+system hosted by [Cancer Genome
+Collaboratory](https://www.cancercollaboratory.org/) and AWS S3.
+
+To rapidly transfer such a large dataset in a well-managed fashion, we
+need a system to orchestrate execution of large number of data transfer
+jobs with the following characteristics:
+
+* simple to develop and use
+* highly reliable
+* able to work with cloud based compute environment
+
+
+## The solution
+
+### Synopsis 
+
+We use JSON files to represent data transfer jobs. Each JSON file contains
+necessary information for a data transfer job. JSON files are
+pre-generated and checked into a GitHub repository, which is structured
+with a few directories each representing a job state. Each data transfer worker clones the GitHub repository and picks up a queued job by moving the corresponding JSON file from 'queued-jobs' to 'downloading-jobs' directory. It then commits the change and pushes back to GitHub central repository. If more than one worker picked up the same job, a Git merge conflict will occur which elegantly avoids double scheduling. As the transfer job progresses to next steps, the worker will add log to the job
+JSON file and move it from one directory to the next, every file change
+and movement will be committed and pushed to GitHub.
+
+### Generate job JSONs
+
+Each job transfers data for a particular GNOS Analysis Object consisting
+of a group of files from a GNOS repository. Job JSON files containing
+information about files to be transfered are generated using metadata
+retrieved from GNOS repositories. Specially designed convention is
+followed to name each job JSON file, which includes multiple parts to
+ensure uniqueness and being informative. For example,
+001a5fa1-dcc8-43e6-8815-fac34eb8a3c9.RECA-EU.C0015.C0015T.WGS-BWA-Tumor.json, it follows this pattern:
```

Comment:
Might want to put some of these in backticks to better highlight it on the page.